### PR TITLE
Introducing an address Registry; Modify contract initialization and address-injection procedure

### PR DESCRIPTION
**Pull request by giladHaimov**
The motivation of this PR, which should be considered as a preliminary step towards a future PR that expands our test-base, is to: 
1. change the way contract-addresses are shared between all contracts 
2. change the contracts initialization flow  
3. removal of the template preprocessing stage thus simplifying both testing and deployment processes and moving to a _single_ codebase for both tests and deployment.

It achieves all that by introducing the following changes into the code;

1. A new address-registry contract that contains all contract-addresses and which is passed to each of the other contracts on construction
2. Moving to constructor-based initialization which removes the need to check contracts  for initialization before actually using them
3. Moving several state-variables from constant to immutable status 
4. Add several internal virtual functions which allows test-logic to be injected by mocks. In several places I have invoked such virtual functions from within the contract's CTOR. Now calling virtual functions from the constructor is usually discouraged, but I have  allowed myself to do so because I was careful not to read the contract's pre-initialized state vars 
within these functions.
5. As part of this effort I have also attempted to standardize the way we transfer funds throughout the contracts.
We will need someday to come up with a policy for dealing with multisig players (i.e.contract not EOA) and accordingly dealing w/ reentancry attacks. Until then I have switched to these 'secure' functions which, at least, verifies that the balance is sufficient before transmitting while preventing reentancry attacks.

I believe that in terms of effect, the removal of the template stage is the most important part of this PR. Template slows us down. We should have zero tolerance for constructs that slow us down.

All this comes with a cost: our successful test count goes down from 166/205 on a clean branch_v.1.1.2r  to 97/205 post changes. This is because some of our tests actually do require conditional-compilation constructs to succeed. I have left these tests so to clarify the cost. Later on we will need to replace them with replicates that do not require conditional compilation (and also add a lot of new tests..)

Additional work was done as part of this PR to generalize eth transfer across all platform contracts so that all of these transfers are now performed using the following System utility functions:
-  _transfer(to,sum) // gas-limited, reverts on error
-  _send(to,sum) // gas-limited, returns false on error
-  _unsafeTransfer(to,sum) // no gas limitation, reverts on error
-  _unsafeSend(to,sum) // no gas limitation, returns false on error

And all of the functions using the _unsafe variants have been added protection against potential reentry attacks

Finally - once this PR is merged, the deployment procedure needs to change so to accommodate these changes, especially with regard to the new Registry contract and CTOR parameters. I have made the necessary changes in the test deployment. We will also need to change mainnet deploy @openissue.

### CODE REVIEW

File: contracts/SystemReward.sol [hunk 1]
```diff
@@ -6,27 +6,24 @@ import "./interface/IParamSubscriber.sol";
...
   uint256 public constant INCENTIVE_BALANCE_CAP = 1e25;
 
   uint256 public incentiveBalanceCap;
-  uint256 public numOperator;
-  mapping(address => bool) operators;
   bool isBurn;
 
   /*********************** init **************************/
-  function init() external onlyNotInit {
-    operators[LIGHT_CLIENT_ADDR] = true;
```
Comment (giladHaimov) [inline contracts/SystemReward.sol:21]: No real need for mapping here

File: contracts/registry/Registry.sol [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        return s_allContracts;
+    }
+
+    function govHubAddr() external view returns (address) {
+        address gotHub = s_allContracts.govHubAddr;
+        assert(gotHub != address(0)); // make sure setAll() was called
+        return gotHub;
+    }
+
+    function verifyAll_(AllContracts memory allContracts) private pure {
```
Comment (giladHaimov) [inline contracts/registry/Registry.sol:38]: assert() is best for denoting non acceptable errors, which is the case here

File: contracts/BtcLightClient.sol [hunk 3]
```diff
@@ -41,10 +39,12 @@ contract BtcLightClient is ILightClient, System, IParamSubscriber{
   uint256 public constant ROUND_SIZE=100;
   uint256 public constant MAXIMUM_WEIGHT=20;
   uint256 public constant CONFIRM_BLOCK = 6;
-  uint256 public constant INIT_ROUND_INTERVAL = 86400;
   uint256 public constant POWER_ROUND_GAP = 7;
   uint256 public constant INIT_STORE_BLOCK_GAS_PRICE = 35e9;
 
+  uint32 public immutable INIT_CHAIN_HEIGHT;
```
Comment (giladHaimov) [inline contracts/BtcLightClient.sol:45]: I prefer upper-casing immutables since, past the constructor, they are effectively constants

File: contracts/System.sol [hunk 4]
```diff
@@ -80,4 +94,104 @@ contract System {
...
+  function _coinbase() internal virtual view returns (address) {
+    return block.coinbase;
+  } 
+
+  function govHub() external returns (address) {
+    return safe_govHubAddr();
+  }
+  // -------
+
+  function secureSend_(address sendTo, uint256 amount) internal returns (bool) {
```
Comment (giladHaimov) [inline contracts/System.sol:113]: We need to come up with a policy for dealing with multisig players (i.e.contract not EOA) and accordingly dealing w/ reentancry attacks. 
Until then I have switched to these 'secure' functions which, at least, verifies that the balance is sufficient before transmitting while preventing reentancry attacks (unlike e.g. Address.sendValue)

File: contracts/System.sol [hunk 5]
```diff
@@ -80,4 +94,104 @@ contract System {
...
+  function sufficientBalance_(uint256 amount) private view returns (bool) {
+    return address(this).balance >= amount;
+  }
+
+
+
+
+  // --- Registry functions below are platform-contracts that are safe to access ---
+
+  function safe_foundationPayable() internal returns(address payable) {
```
Comment (giladHaimov) [inline contracts/System.sol:134]: All platform contracts are 'safe_' in the sense that we can rely on their trusted behavior with no additional checks/protection (Reentrry, DoS, the lot..).  Other addresses should at least be treated - I would suggest also prefixed - as 'unsafe'

File: contracts/CandidateHub.sol [hunk 6]
```diff
@@ -86,16 +87,19 @@ contract CandidateHub is ICandidateHub, System, IParamSubscriber {
...
-    validatorCount = INIT_VALIDATOR_COUNT;
+    roundInterval = roundInterval_;
+    validatorCount = validatorCount_;
     maxCommissionChange = MAX_COMMISSION_CHANGE;
     roundTag = 7;
-    alreadyInit = true;
   }
-  
+
+  function setControlRoundTimeTag(bool value) external onlyGov {
```
Comment (giladHaimov) [inline contracts/CandidateHub.sol:99]: This setter is for testing only and should never be called from production code. Since marked as onlyGov the additional risk is contained, still I would remove it from our code as soon as we can

File: contracts/GovHub.sol [hunk 7]
```diff
@@ -89,18 +88,21 @@ contract GovHub is System, IParamSubscriber {
...
 
-  function init() external onlyNotInit {
+  constructor(Registry registry, uint256 votingPeriod_, uint256 executingPeriod_, bytes memory initialMembers_) System(registry) {
     proposalMaxOperations = PROPOSAL_MAX_OPERATIONS;
-    votingPeriod = VOTING_PERIOD;
-    executingPeriod = EXECUTING_PERIOD;
-    RLPDecode.RLPItem[] memory items = INIT_MEMBERS.toRLPItem().toList();
+    votingPeriod = votingPeriod_;
+    executingPeriod = executingPeriod_;
+    _loadInitialMembers(initialMembers_);
```
Comment (giladHaimov) [inline contracts/GovHub.sol:95]: Calling virtual functions from the constructor is usually discouraged. I have allowed myself to do so here because in _loadInitialMembers is only used to populate with initial members and in no place does it attempt to read contract's pre-initialized state vars

File: contracts/PledgeAgent.sol [hunk 8]
```diff
@@ -295,7 +298,7 @@ contract PledgeAgent is IPledgeAgent, System, IParamSubscriber {
   /// @param amount The amount of CORE to undelegate
   function undelegateCoin(address agent, uint256 amount) public {
     (uint256 deposit, ) = undelegateCoin(agent, msg.sender, amount, false);
-    Address.sendValue(payable(msg.sender), deposit);
```
Comment (giladHaimov) [inline contracts/PledgeAgent.sol:298]: Avoiding potential reentrancy attack

File: contracts/PledgeAgent.sol [hunk 9]
```diff
@@ -366,7 +369,7 @@ contract PledgeAgent is IPledgeAgent, System, IParamSubscriber {
 
   /*********************** Internal methods ***************************/
   function distributeReward(address payable delegator, uint256 reward) internal {
-    Address.sendValue(delegator, reward);
+    secureTransfer_(delegator, reward);
```
Comment (giladHaimov) [inline contracts/PledgeAgent.sol:372]: Avoiding potential reentrancy attack

File: contracts/System.sol [hunk 10]
```diff
@@ -1,71 +1,85 @@
...
-    
-    require(tx.gasprice == 0 , "gasprice is not zero");
-    
+    require(_gasprice() == 0 , "gasprice is not zero");
     _;
   }
 
-  modifier onlyNotInit() {
-    require(!alreadyInit, "the contract already init");
+  modifier onlyNotInit() { //placeholder
```
Comment (giladHaimov) [inline contracts/System.sol:51]: init() mechanism is no longer used

File: contracts/System.sol [hunk 11]
```diff
@@ -80,4 +94,104 @@ contract System {
...
+    return s_allContracts.slashIndicator;
+  }  
+
+  function safe_govHubAddr() internal returns(address) {
+    _cacheAllContractsLocally();
+    return s_allContracts.govHubAddr;
+  }
+
+  function _cacheAllContractsLocally() private {
+    //@correlate-registry.cache
```
Comment (giladHaimov) [inline contracts/System.sol:191]: allContracts may only be cached if immutable per contracts' lifecycle. See Registry.setAll()

File: contracts/ValidatorSet.sol [hunk 12]
```diff
@@ -66,17 +65,20 @@ contract ValidatorSet is IValidatorSet, System, IParamSubscriber {
...
   /*********************** init **************************/
-  function init() external onlyNotInit {
-    (Validator[] memory validatorSet, bool valid) = decodeValidatorSet(INIT_VALIDATORSET_BYTES);
+  constructor(Registry registry, bytes memory validatorBytes_) System(registry) {
+    blockReward = BLOCK_REWARD;
+    blockRewardIncentivePercent = BLOCK_REWARD_INCENTIVE_PERCENT;
+    _loadInitialValidatorSet(validatorBytes_);
+  }
+
+  function _loadInitialValidatorSet(bytes memory validatorBytes_) internal virtual {
```
Comment (giladHaimov) [inline contracts/ValidatorSet.sol:74]: See my comment above regarding accessing virtual functions from CTOR

File: contracts/test/pledgeAgentProxy.sol [hunk 13]
```diff
@@ -6,7 +7,7 @@ contract PledgeAgentProxy {
 
     event delegate(bool success);
     event claim(uint256 reward, bool allClaimed);
-    constructor(address pledgeAgentAddress) public {
```
Comment (giladHaimov) [inline contracts/test/pledgeAgentProxy.sol:9]: ignored by solc

File: contracts/utils/ImmutableOwner.sol [hunk 14]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: Apache2.0
+pragma solidity 0.8.4;
+
+abstract contract ImmutableOwner {
+    address public immutable s_owner = msg.sender;
```
Comment (giladHaimov) [inline contracts/utils/ImmutableOwner.sol:5]: used to validate that the registry is set only by the contract's deployer

File: contracts/System.sol [hunk 15]
```diff
@@ -80,4 +94,104 @@ contract System {
...
+  function _coinbase() internal virtual view returns (address) {
+    return block.coinbase;
+  } 
+
+  function govHub() external returns (address) {
+    return safe_govHubAddr();
+  }
+  // -------
+
+  function _secureSend(address sendTo, uint256 amount) internal returns (bool) {
```
Comment (denalimarsh) [inline contracts/System.sol:113]: This standardization will re-introduce the GnosisSafe multisig transfer issue, please consider and restore originaly functionality for the relevant `transfer`s

File: contracts/System.sol [hunk 16]
```diff
@@ -1,71 +1,85 @@
...
-  modifier onlyInit() {
-    require(alreadyInit, "the contract not init yet");
+  modifier onlyInit() { //placeholder
+    //require(s_alreadyInit, "the contract not init yet");
     _;
   }
 
   modifier onlySlash() {
-    require(msg.sender == SLASH_CONTRACT_ADDR, "the msg sender must be slash contract");
+    require(msg.sender == address(safe_slashIndicator()), "the msg sender must be slash contract");
```
Comment (denalimarsh) [inline contracts/System.sol:62]: As discussed let's drop the `_safe` prefix for system contracts, instead denoting non-platform contracts as `_unsafe`
```suggestion
    require(msg.sender == address(slashIndicator()), "the msg sender must be slash contract");
```

**Review by denalimarsh — CHANGES_REQUESTED** [review]
Initial review, will do another pass after the changes we discussed.

We're introducing two new tags:
- `@OpenIssue`: something that needs to be resolved or an area of code with potential implications we shouldn't forget about. Before releasing a new mainnet version, all `@OpenIssue` tags must be resolved and removed from the codebase.
- `@correlate`: calls out a non-obvious dependency relationship between two areas of code. If changing code with a `@correlate` tag, you need to look at the area of the code it is referencing. Correlate serves as a prefix to a unique relationship identifier e.g. `@correlate-registry.cache` which can be searched to find all relevant code areas.

Some of the new codebase standards we aligned on:
- Use immutable variables instead of constant variables where appropriate
- Prefix state/storage variables with `s_`. If creating a storage reference a state variable, prefix it with `sref_`
- Prefix internal functions with `_`
- Prefix non-platform contracts (contracts outside this repo) with `_unsafe`

File: contracts/PledgeAgent.sol [hunk 17]
```diff
@@ -118,11 +121,11 @@ contract PledgeAgent is IPledgeAgent, System, IParamSubscriber {
   /// @param target Address of the target candidate
   error SameCandidate(address source, address target);
 
-  function init() external onlyNotInit {
+  constructor(Registry registry, uint256 powerBlockFactor_) System(registry) { //@openissue on mainnet: powerBlockFactor_ = 1e18
```
[the remark this answers is not in the document]
Comment (giladHaimov) [reply contracts/PledgeAgent.sol:124]: @openissue indicates a location in the code that needs to be resolved.
Resolving may involve an investigation that determines that there's no real issue here, in which case @openissue may be replaced with a comment. It may also determine that a fix is needed, in which case the code should be changed.
Regardless - before deploying onto mainnet the @openissue count should be zero

File: contracts/System.sol (hunk 15 above)
Comment (giladHaimov) [reply contracts/System.sol:113]: As discussed I will revert all calls with no gas limitation, which allows for multisigs clients, to how they were before

File: contracts/CandidateHub.sol [hunk 18]
```diff
@@ -2,23 +2,23 @@
 pragma solidity 0.8.4;
 import "./lib/BytesToTypes.sol";
 import "./lib/Memory.sol";
+import {ReentrancyGuard} from "./lib/ReentrancyGuard.sol";
```
Comment (giladHaimov) [inline contracts/CandidateHub.sol:5]: I am aware of the fact that the relevant functions are structured as checks-effects-interactions. I'm still adding a guard as a 2nd layer protection against future (unsafe) changes

File: contracts/GovHub.sol [hunk 19]
```diff
@@ -208,29 +211,34 @@ contract GovHub is System, IParamSubscriber {
...
+        callData = proposal.calldatas[i]; 
       } else {
-        callData = abi.encodePacked(bytes4(keccak256(bytes(proposal.signatures[i]))), proposal.calldatas[i]);
+        // function call
+        bytes4 functionSelector = bytes4(keccak256(bytes(proposal.signatures[i])));
+        callData = abi.encodePacked(functionSelector, proposal.calldatas[i]);
       }
 
-      (bool success, bytes memory returnData) = proposal.targets[i].call{ value: proposal.values[i] }(callData);
+      //@openissue: is it guaranteed that all proposal targets will be platform contracts?
```
Comment (denalimarsh) [inline contracts/GovHub.sol:230]: I think it's reasonable to assume that all proposal targets will be platform contracts. I can't think of any cases where we'd want governance to control a parameter outside of the set of platform contracts. Can keep as @openissue for now.

File: contracts/PledgeAgent.sol [hunk 20]
```diff
@@ -118,11 +121,12 @@ contract PledgeAgent is IPledgeAgent, System, IParamSubscriber {
   /// @param target Address of the target candidate
   error SameCandidate(address source, address target);
 
-  function init() external onlyNotInit {
+  constructor(Registry registry, uint256 powerBlockFactor_) System(registry) { 
+    //@openissue:deploy : note that on mainnet initial powerBlockFactor_ shoiuld be = 1e18
```
Comment (denalimarsh) [inline contracts/PledgeAgent.sol:125]: Nit
```suggestion
    //@openissue:deploy : note that on mainnet initial powerBlockFactor_ should be = 1e18
```

File: contracts/PledgeAgent.sol (hunk 8 above)
Comment (denalimarsh) [reply contracts/PledgeAgent.sol:298]: In the future testing PRs let's write some malicious contracts that attempt reentrancy exploits

File: contracts/System.sol (hunk 10 above)
Comment (denalimarsh) [reply contracts/System.sol:52]: add `@openissue`, would like to see this removed if possible before a mainnet deploy

File: contracts/System.sol [hunk 21]
```diff
@@ -1,71 +1,86 @@
...
-  modifier onlyNotInit() {
-    require(!alreadyInit, "the contract already init");
+  modifier onlyNotInit() { //placeholder
+    //require(!s_alreadyInit, "the contract already init");
     _;
   }
 
-  modifier onlyInit() {
-    require(alreadyInit, "the contract not init yet");
+  modifier onlyInit() { //placeholder
```
Comment (denalimarsh) [inline contracts/System.sol:57]: same here

File: contracts/System.sol [hunk 22]
```diff
@@ -1,71 +1,86 @@
...
+    //require(s_alreadyInit, "the contract not init yet");
     _;
   }
 
   modifier onlySlash() {
-    require(msg.sender == SLASH_CONTRACT_ADDR, "the msg sender must be slash contract");
+    require(msg.sender == address(_slashIndicator()), "the msg sender must be slash contract");
```
Comment (denalimarsh) [inline contracts/System.sol:63]: let's be exact in the error messages
```suggestion
    require(msg.sender == address(_slashIndicator()), "the msg sender must be slashIndicator contract");
```

File: contracts/System.sol [hunk 23]
```diff
@@ -1,71 +1,86 @@
...
     _;
   }
 
   modifier onlyGov() {
-    require(msg.sender == GOV_HUB_ADDR, "the msg sender must be governance contract");
+    require(msg.sender == _govHubAddr(), "the msg sender must be governance contract");
```
Comment (denalimarsh) [inline contracts/System.sol:68]:
```suggestion
    require(msg.sender == _govHubAddr(), "the msg sender must be govHub contract");
```

File: contracts/System.sol [hunk 24]
```diff
@@ -1,71 +1,86 @@
...
     _;
   }
 
   modifier onlyCandidate() {
-    require(msg.sender == CANDIDATE_HUB_ADDR, "the msg sender must be candidate contract");
+    require(msg.sender == address(_candidateHub()), "the msg sender must be candidate contract");
```
Comment (denalimarsh) [inline contracts/System.sol:73]:
```suggestion
    require(msg.sender == address(_candidateHub()), "the msg sender must be candidateHub contract");
```

File: contracts/System.sol (hunk 22 above)
Comment (denalimarsh) [reply contracts/System.sol:63]: It doesn't read as nicely, but good if there's ever any confusion about which contract the error message is referring to

File: contracts/System.sol [hunk 25]
```diff
@@ -80,4 +95,118 @@ contract System {
...
+  }
+
+  function _coinbase() internal virtual view returns (address) {
+    return block.coinbase;
+  } 
+
+  function govHub() external returns (address) {
+    return _govHubAddr();
+  }
+  // -------
```
Comment (denalimarsh) [inline contracts/System.sol:112]: Delete or add comment text describing the section
```suggestion
```

File: contracts/System.sol [hunk 26]
```diff
@@ -80,4 +95,118 @@ contract System {
...
+    Address.sendValue(payable(sendTo), amount);
+  }
+
+  function _sufficientBalance(uint256 amount) private view returns (bool) {
+    return address(this).balance >= amount;
+  }
+
+
+
+
```
Comment (denalimarsh) [inline contracts/System.sol:146]:
```suggestion
```

File: contracts/System.sol [hunk 27]
```diff
@@ -80,4 +95,118 @@ contract System {
...
+  }  
+
+  function _govHubAddr() internal returns(address) {
+    _cacheAllContractsLocally();
+    return s_allContracts.govHubAddr;
+  }
+
+  function _cacheAllContractsLocally() private {
+    //@correlate-registry.cache
+    if (s_allContracts.burn == IBurn(address(0))) {
```
Comment (denalimarsh) [inline contracts/System.sol:207]: Should be fine. Had a thought though: what if a future CoreSDK blockchain forks this codebase and removes the `Burn` contract because they don't want the burn functionality? I guess it's reasonable for it to be on them to resolve this, as any other approach results in us expensively checking a lot more storage.

File: contracts/SystemReward.sol [hunk 28]
```diff
@@ -104,4 +100,8 @@ contract SystemReward is System, ISystemReward, IParamSubscriber {
     }
     emit paramChange(key, value);
   }
+
+  function _isOperator(address addr) internal virtual returns (bool) {
```
Comment (denalimarsh) [inline contracts/SystemReward.sol:104]: Much cleaner!

File: contracts/mock/GovHubMock.sol [hunk 29]
```diff
@@ -1,8 +1,16 @@
...
 pragma solidity 0.8.4;
 
 import "../GovHub.sol";
+import "../registry/Registry.sol";
 
 contract GovHubMock is GovHub {
+    constructor(Registry registry, uint256 votingPeriod_, uint256 executingPeriod_, bytes memory membersBytes_) 
+            GovHub(registry, votingPeriod_, executingPeriod_, membersBytes_) {}
+        
+
```
Comment (denalimarsh) [inline contracts/mock/GovHubMock.sol:11]:
```suggestion
```

File: contracts/mock/ValidatorSetMock.sol [hunk 30]
```diff
@@ -1,10 +1,41 @@
...
+    }
+
+    function _gasprice() internal override view returns (uint) {
+      return txgasprice;
+    }
+
+    function _coinbase() internal override view returns (address) {
+      return blockcoinbase;
+    }  
+
```
Comment (denalimarsh) [inline contracts/mock/ValidatorSetMock.sol:32]:
```suggestion
```

File: contracts/registry/AllContracts.sol [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+struct AllContracts {
+    IBurn burn;
+    ILightClient lightClient;
+    ISlashIndicator slashIndicator;
+    ISystemReward systemReward;
+    ICandidateHub candidateHub;
+    IPledgeAgent pledgeAgent;        
+    IValidatorSet validatorSet;
+    IRelayerHub relayerHub;
+    address foundationAddr;
```
Comment (denalimarsh) [inline contracts/registry/AllContracts.sol:22]: Just calling out that `foundationAddr` is the address that we can expect to be set to a multisig

File: contracts/utils/SetOnce.sol [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+    modifier setOnlyOnce() {
+        require(!s_wasSet, "already set");
+        _markAsSet();
+        _;
+    }
+
+    function _markAsSet() internal virtual {
+        s_wasSet = true;
+    }
+
```
Comment (denalimarsh) [inline contracts/utils/SetOnce.sol:16]:
```suggestion
```

**Review by denalimarsh — APPROVED** [review]
LGTM, but now that we're breaking tests we'll need to fix them up and return to pre-existing test parity in the next PR at a minimum.

**Comment by giladHaimov** [comment]
> Initial review, will do another pass after the changes we discussed.
> 
> We're introducing two new tags:
> 
> * `@OpenIssue`: something that needs to be resolved or an area of code with potential implications we shouldn't forget about. Before releasing a new mainnet version, all `@OpenIssue` tags must be resolved and removed from the codebase.
> * `@correlate`: calls out a non-obvious dependency relationship between two areas of code. If changing code with a `@correlate` tag, you need to look at the area of the code it is referencing. Correlate serves as a prefix to a unique relationship identifier e.g. `@correlate-registry.cache` which can be searched to find all relevant code areas.
> 
> Some of the new codebase standards we aligned on:
> 
> * Use immutable variables instead of constant variables where appropriate
> * Prefix state/storage variables with `s_`. If creating a storage reference a state variable, prefix it with `sref_`
> * Prefix internal functions with `_`
> * Prefix non-platform contracts (contracts outside this repo) with `_unsafe`

yep agreed. it will probably be a good idea to issue a code-standards/best-practices draft for the entire team to review.

File: contracts/GovHub.sol (hunk 19 above)
Comment (giladHaimov) [reply contracts/GovHub.sol:230]: If that is the case we must assert it as part of a proposal acceptance!

File: contracts/registry/AllContracts.sol (hunk 31 above)
Comment (giladHaimov) [reply contracts/registry/AllContracts.sol:22]: foundationAddr is only used receiveRewards() for secure eth transfer so we're probably good here

File: contracts/System.sol (hunk 22 above)
Comment (giladHaimov) [reply contracts/System.sol:63]: agreed

**Comment by giladHaimov** [comment]
To be replaced with a new PR that avoids any changes of the golang engine plus takes care to keep existing storage layout - thus not breaking the update mechanism
